File: DSAlgorithms/patterns/sliding_window/12_word_concat.ts
const word_concat = (str: string, words: string[]) => {
    if(words.length === 0 || words[0].length === 0)
        return [];
    const word_freq = new Map<string, number>();
    words.forEach((word) => {
        if(!word_freq.has(word))
            word_freq.set(word, 1);
        else
        {
            const current = word_freq.get(word);
            if(current !== undefined)
                word_freq.set(word, current + 1);
        }
    });
    const result_indices: number[] = [];
    const words_count = words.length;
    const word_length = words[0].length;
    for(let i = 0; i < (str.length - words_count * word_length) + 1; i++)
    {
        const words_seen = new Map<string, number>();
        for(let j = 0; i < words_count; j++)
        {
            let next_word_index = i + j * word_length;
            let word = str.substring(next_word_index, next_word_index + word_length);
            if(!word_freq.has(word))
                break;
            if(!words_seen.has(word))
                words_seen.set(word, 1);
            else
            {
                const current = words_seen.get(word);
                if(current !== undefined)
                    words_seen.set(word, current + 1);
            }
            const a = words_seen.get(word);
            const b = word_freq.get(word);
            if(a !== undefined && b !== undefined )
                if(a > (b || 0))
                    break;
            if(j + 1 === words_count)
                result_indices.push(i);
        }
    }
    return result_indices;
}
console.log(word_concat('catfoxcat', ['cat', 'fox']));
console.log(word_concat('catcatfoxfox', ['cat', 'fox']));

File: DSAlgorithms/patterns/sliding_window/12_word_concat.py
def find_concat(str, words):
    if len(words) == 0 or len(words[0]) == 0:
        return []
    word_frequency = {}
    for word in words:
        if word not in word_frequency:
            word_frequency[word] = 0
        word_frequency[word] += 1
    result_indices = []
    words_count = len(words)
    word_length = len(words[0])
    for i in range((len(str) - words_count * word_length) + 1):
        words_seen = {}
        for j in range(0, words_count):
            next_word_index = i + j * word_length
            word = str[next_word_index: next_word_index + word_length]
            if word not in word_frequency:
                break
            if word not in words_seen:
                words_seen[word] = 0
            words_seen[word] += 1
            if words_seen[word] > word_frequency.get(word, 0):
                break;
            if j + 1 == words_count:
                result_indices.append(i)
    return result_indices

def main():
    print(find_concat("catfoxcat", ["cat", "fox"]))
    print(find_concat("catcatfoxfox", ["cat", "fox"]))

main()
